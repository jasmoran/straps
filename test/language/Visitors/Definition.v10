#!Definition.out
#-

include "src/stdlib/stdlib.v10"
include "src/language/parser/Parser.v10"
include "src/language/Scope.v10"
include "src/language/Type.v10"

include "src/language/visitors/Visitor.v10"
include "src/language/visitors/Printer.v10"
include "src/language/visitors/Constructor.v10"
include "src/language/visitors/Typing.v10"
include "src/language/visitors/Method.v10"
include "src/language/visitors/Definition.v10"

global(Scope scope)

func void main()(Stream input, Parser parser, Node root, Visitor v, Scope sc) {
	# Open input file
	input = File_open("test/language/Visitors/sample.v10.test")
	if (input < 0) {
		"Couldn't read input file".println(STDERR)
		return 1
	}

	parser = Parser_new("sample.v10", input)
	root = parser.parse()
	sc = Scope_new()
	scope = sc

	v = Constructor_new()
	root.accept(v)
	v = Typing_new()
	root.accept(v)
	v = Method_new()
	root.accept(v)
	v = Definition_new()
	root.accept(v)
	v = Printer_new()
	root.accept(v)

	sc.print()
}
