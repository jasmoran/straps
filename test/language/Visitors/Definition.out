nBLOCK
  nCLASS(Any)
    nLIST
    nLIST
  nCLASS(Character)
    nLIST
    nLIST
  nCLASS(Number)
    nLIST
    nLIST
  nCLASS(Boolean)
    nLIST
    nLIST
  nENUM(Boolean)
    nBLOCK
      nIDENTIFIER(Unknown: "false")
      nIDENTIFIER(Unknown: "true")
  nCLASS(String)
    nLIST
    nLIST
  nFUNCTION(#Any#: allocate_Number)
size: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
    nLIST
      nIDENTIFIER(#Number#: "size")
    nLIST
    nBLOCK
  nFUNCTION(#Boolean#: thisisafunction_Any_Any)
arg1: <<
  nBINARYOP(#Any#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
arg2: <<
  nBINARYOP(#Any#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
vrbl1: <<
  nBINARYOP(#Any#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
vrbl2: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-16")
>>
    nLIST
      nIDENTIFIER(#Any#: "arg1")
      nIDENTIFIER(#Any#: "arg2")
    nLIST
      nIDENTIFIER(#Any#: "vrbl1")
      nIDENTIFIER(#Number#: "vrbl2")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nIDENTIFIER(Unknown: "arg1")
        nIDENTIFIER(Unknown: "vrbl1")
      nUNARYOP(Unknown: INCREMENT, right)
        nIDENTIFIER(Unknown: "vrbl2")
      nBINARYOP(Unknown: ADD)
        nIDENTIFIER(Unknown: "vrbl1")
        nNUMBER(#Number#: "1337")
      nBINARYOP(Unknown: DOT)
        nBINARYOP(Unknown: PAREN_OPEN)
          nBINARYOP(Unknown: DOT)
            nBINARYOP(Unknown: PAREN_OPEN)
              nBINARYOP(Unknown: DOT)
                nBINARYOP(Unknown: DOT)
                  nBINARYOP(Unknown: DOT)
                    nIDENTIFIER(Unknown: "true")
                    nIDENTIFIER(Unknown: "as_string")
                  nIDENTIFIER(Unknown: "random_char")
                nIDENTIFIER(Unknown: "toString")
              nCALL
            nIDENTIFIER(Unknown: "println")
          nCALL
            nNUMBER(#Number#: "77")
            nIDENTIFIER(Unknown: "vrbl2")
        nIDENTIFIER(Unknown: "deref")
      nLOOP
        nBLOCK
          nBINARYOP(Unknown: PAREN_OPEN)
            nBINARYOP(Unknown: DOT)
              nSTRING(#String#: "oopsy daisy")
              nIDENTIFIER(Unknown: "println")
            nCALL
              nNUMBER(#Number#: "1")
              nNUMBER(#Number#: "2")
      nRETURN
        nIDENTIFIER(Unknown: "true")
  nFUNCTION(#Any#: Any_new_Any)
deref: <<
  nBINARYOP(#Any#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(#Any#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
      nIDENTIFIER(#Any#: "deref")
    nLIST
      nIDENTIFIER(#Any#: "obj")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nIDENTIFIER(#Any#: "obj")
        nBINARYOP(Unknown: PAREN_OPEN)
          nIDENTIFIER(#Any#: "allocate")
          nCALL
            nNUMBER(#Number#: "8")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(Unknown: DOT)
          nIDENTIFIER(#Any#: "obj")
          nIDENTIFIER(#Any#: "deref")
        nIDENTIFIER(#Any#: "deref")
      nRETURN
        nIDENTIFIER(#Any#: "obj")
  nFUNCTION(#Character#: Character_new)
obj: <<
  nBINARYOP(#Character#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
      nIDENTIFIER(#Character#: "obj")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nIDENTIFIER(#Character#: "obj")
        nBINARYOP(Unknown: PAREN_OPEN)
          nIDENTIFIER(#Character#: "allocate")
          nCALL
            nNUMBER(#Number#: "0")
      nRETURN
        nIDENTIFIER(#Character#: "obj")
  nFUNCTION(#Number#: Number_new)
obj: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
      nIDENTIFIER(#Number#: "obj")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nIDENTIFIER(#Number#: "obj")
        nBINARYOP(Unknown: PAREN_OPEN)
          nIDENTIFIER(#Number#: "allocate")
          nCALL
            nNUMBER(#Number#: "0")
      nRETURN
        nIDENTIFIER(#Number#: "obj")
  nFUNCTION(#Boolean#: Boolean_new_String)
as_string: <<
  nBINARYOP(#String#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(#Boolean#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
      nIDENTIFIER(#String#: "as_string")
    nLIST
      nIDENTIFIER(#Boolean#: "obj")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nIDENTIFIER(#Boolean#: "obj")
        nBINARYOP(Unknown: PAREN_OPEN)
          nIDENTIFIER(#Boolean#: "allocate")
          nCALL
            nNUMBER(#Number#: "8")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(Unknown: DOT)
          nIDENTIFIER(#Boolean#: "obj")
          nIDENTIFIER(#String#: "as_string")
        nIDENTIFIER(#String#: "as_string")
      nRETURN
        nIDENTIFIER(#Boolean#: "obj")
  nFUNCTION(#String#: String_new_Character)
random_char: <<
  nBINARYOP(#Character#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(#String#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
      nIDENTIFIER(#Character#: "random_char")
    nLIST
      nIDENTIFIER(#String#: "obj")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nIDENTIFIER(#String#: "obj")
        nBINARYOP(Unknown: PAREN_OPEN)
          nIDENTIFIER(#String#: "allocate")
          nCALL
            nNUMBER(#Number#: "8")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(Unknown: DOT)
          nIDENTIFIER(#String#: "obj")
          nIDENTIFIER(#Character#: "random_char")
        nIDENTIFIER(#Character#: "random_char")
      nRETURN
        nIDENTIFIER(#String#: "obj")
  nFUNCTION(#String#: _Character_toString_Character)
this: <<
  nBINARYOP(#Character#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
    nLIST
      nIDENTIFIER(#Character#: "this")
    nLIST
    nBLOCK
  nFUNCTION(#Any#: _String_println_String_Number_Number)
this: <<
  nBINARYOP(#String#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
stream: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
ilikenumbers: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "32")
>>
    nLIST
      nIDENTIFIER(#String#: "this")
      nIDENTIFIER(#Number#: "stream")
      nIDENTIFIER(#Number#: "ilikenumbers")
    nLIST
    nBLOCK
Any: <<
  nCLASS(Any)
deref: <<
  nINDEX(#Any#: "0")
>>
    nLIST
    nBLOCK
>>
Character: <<
  nCLASS(Character)
toString: <<
  nFUNCTION(#String#: _Character_toString_Character)
this: <<
  nBINARYOP(#Character#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
    nLIST
    nLIST
    nBLOCK
>>
    nLIST
    nBLOCK
>>
Number: <<
  nCLASS(Number)
    nLIST
    nBLOCK
>>
Boolean: <<
  nCLASS(Boolean)
as_string: <<
  nINDEX(#String#: "0")
>>
    nLIST
    nBLOCK
>>
String: <<
  nCLASS(String)
println_Number_Number: <<
  nFUNCTION(#Any#: _String_println_String_Number_Number)
this: <<
  nBINARYOP(#String#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
stream: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
ilikenumbers: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "32")
>>
    nLIST
    nLIST
    nBLOCK
>>
random_char: <<
  nINDEX(#Character#: "0")
>>
    nLIST
    nBLOCK
>>
false: <<
  nNUMBER(#Boolean#: "0")
>>
true: <<
  nNUMBER(#Boolean#: "1")
>>
allocate_Number: <<
  nFUNCTION(#Any#: allocate_Number)
size: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
    nLIST
    nLIST
    nBLOCK
>>
thisisafunction_Any_Any: <<
  nFUNCTION(#Boolean#: thisisafunction_Any_Any)
arg1: <<
  nBINARYOP(#Any#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
arg2: <<
  nBINARYOP(#Any#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
vrbl1: <<
  nBINARYOP(#Any#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
vrbl2: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-16")
>>
    nLIST
    nLIST
    nBLOCK
>>
Any_new_Any: <<
  nFUNCTION(#Any#: Any_new_Any)
deref: <<
  nBINARYOP(#Any#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(#Any#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
    nBLOCK
>>
Character_new: <<
  nFUNCTION(#Character#: Character_new)
obj: <<
  nBINARYOP(#Character#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
    nBLOCK
>>
Number_new: <<
  nFUNCTION(#Number#: Number_new)
obj: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
    nBLOCK
>>
Boolean_new_String: <<
  nFUNCTION(#Boolean#: Boolean_new_String)
as_string: <<
  nBINARYOP(#String#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(#Boolean#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
    nBLOCK
>>
String_new_Character: <<
  nFUNCTION(#String#: String_new_Character)
random_char: <<
  nBINARYOP(#Character#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
obj: <<
  nBINARYOP(#String#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "-8")
>>
    nLIST
    nLIST
    nBLOCK
>>
_Character_toString_Character: <<
  nFUNCTION(#String#: _Character_toString_Character)
this: <<
  nBINARYOP(#Character#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
    nLIST
    nLIST
    nBLOCK
>>
_String_println_String_Number_Number: <<
  nFUNCTION(#Any#: _String_println_String_Number_Number)
this: <<
  nBINARYOP(#String#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "16")
>>
stream: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "24")
>>
ilikenumbers: <<
  nBINARYOP(#Number#: SQUARE_OPEN)
    nREGISTER(Unknown: "rbp")
    nNUMBER(#Number#: "32")
>>
    nLIST
    nLIST
    nBLOCK
>>
